### ISSUE TITLE
Replace `()` error types with specific types that implement Error

### ISSUE THREAD

**Issue by brson** [servo/rust-url#299]
There are a number of methods in url that return `Result<_, ()>`. `()` though does not implement `Error` and so does not inter operate cleanly with callers that want to treat the result as `Error` (like error-chain).

In today's Rust the best pattern for this is probably to create a single `Error` enum for the entire crate to share and return it everywhere.

Would require a major version bump.

**Comment by dtolnay** [comment]
Yep we struggled with this in #323. It makes it hard to handle these errors concisely.

**Comment by Enet4** [comment]
Here's a list of public methods returning `()` as an error:

1.  [`Url#path_segments_mut`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1038)
1. [`Url#set_port`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1089)
1. [`Url#set_ip_host`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1258)
1. [`Url#set_password`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1274)
1. [`Url#set_username`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1323)
1. [`Url#set_scheme`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1410)
1. [`Url#from_file_path`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1460)
1. [`Url#from_directory_path`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1495)
1. [`Url#to_file_path`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L1575)

Although [`Url#path_segments`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L854) returns an `Option`, this is not very consistent when we look at its "_mut" counterpart `path_segments_mut`, which returns a `Result`.

I would have suggested a `CannotBeBase` variant of `ParseError` for all methods requiring a URL that is not cannot-be-base (this includes methods 1 to 6 above), but it seems that the [current design](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/parser.rs#L55-L56) is to have one for each method. Should we change that to make the error enum more concise? I cannot think of a good reason to distinguish `SetHostCannotBeBase` from `PathSegmentsCannotBeBase` or `SetSchemeCannotBeBase`.

Let's not also forget [`Url#with_default_port`](https://github.com/servo/rust-url/blob/780410c0f161f57bc1a58e40fab1930b703456da/src/lib.rs#L799), which takes an argument of type `FnOnce() -> Result<u16, ()>`. Once we harmonize all the results, we might be able to just replace the error type with `ParseError`.

**Comment by tmccombs** [comment]
I'll start working on this.

**Comment by nox** [comment]
> I would have suggested a `CannotBeBase` variant of `ParseError` for all methods requiring a URL that is not cannot-be-base (this includes methods 1 to 6 above)

1 to 6 don't exactly have the same error conditions, actually.

* 2 wants an non-`file:` URL with a host.
* 4-6 want a URL with a host.

**Comment by nox** [comment]
I guess for 4-6 we could return `EmptyHost`, but I am a bit weirded out by the `self.host() == Some(Host::Domain(""))` test in `set_password` which is not found in the others.

**Comment by tmccombs** [comment]
In my PR I went with `EmptyHost` for 4-6. But would also be ok with making a new error type for them.

**Comment by bstrie** [comment]
@tmccombs Are you pursuing this? If not, I might be interested in resurrecting your patch.

**Comment by tmccombs** [comment]
tbh, I'm not entirely sure what to do with this, since I missed the last major release and this is a backwards incompatible change.

**Comment by bstrie** [comment]
@tmccombs I have an idea for a way to make this happen that wouldn't involve a breaking change, I'll take a crack at it based on your PR.

**Comment by djc** [comment]
As I understand it, this change would be semver-incompatible. I've opened a tracking issue for proposed semver-incompatible changes to discuss how these could be handled going forward.

**Comment by mhnap** [comment]
Maybe there are some updates on this issue? Or some help is needed to fix it?
